gridstatus/caiso.py

```diff
@@ -236,16 +242,31 @@ def get_solar_and_wind_forecast_dam(self, date, end=None, verbose=False):
             raw_data=False,
         )
 
+        # Day-ahead hourly wind and solar forecast is published at 7:00 AM according
+        # to OASIS.
+        data["Publish Time"] = (
+            data["Time"].dt.floor("D") + pd.Timedelta(hours=7) - pd.Timedelta(days=1)
```

Comment:
the publish time isn't correct for further out days using this method

<img width="1181" alt="image" src="https://github.com/gridstatus/gs-etl/assets/371792/7692ba67-ea07-4325-a3f7-31286efe8a0f">

im not see any obvious way in the api response or documentation to figure out publish time

https://www.caiso.com/Documents/OASIS-InterfaceSpecification_v5_1_1Clean_Fall2017Release.pdf

Reply:
Hmm, I think we can do something like:

* For past days through today, the publish time is as written
* For tomorrow and future days, the publish time is today at 7 am

Reply:
The same problem (publish times in the future), currently applies to `get_load_forecast`:

```python
CAISO().get_load_forecast('2024-04-10', end='2024-04-12')

| Publish Time                      Time            Interval Start              Interval End  Load Forecast                 | 
|---------------------------------------------------------------------------------------------------------------------------| 
| 0  2024-04-09 09:10:00-07:00 2024-04-10 00:00:00-07:00 2024-04-10 00:00:00-07:00 2024-04-10 01:00:00-07:00    20_663.9300 | 
| 1  2024-04-09 09:10:00-07:00 2024-04-10 01:00:00-07:00 2024-04-10 01:00:00-07:00 2024-04-10 02:00:00-07:00    19_970.4700 | 
| 2  2024-04-09 09:10:00-07:00 2024-04-10 02:00:00-07:00 2024-04-10 02:00:00-07:00 2024-04-10 03:00:00-07:00    19_482.9900 | 
| 3  2024-04-09 09:10:00-07:00 2024-04-10 03:00:00-07:00 2024-04-10 03:00:00-07:00 2024-04-10 04:00:00-07:00    19_236.4100 | 
| 4  2024-04-09 09:10:00-07:00 2024-04-10 04:00:00-07:00 2024-04-10 04:00:00-07:00 2024-04-10 05:00:00-07:00    19_681.5300 | 
| 5  2024-04-09 09:10:00-07:00 2024-04-10 05:00:00-07:00 2024-04-10 05:00:00-07:00 2024-04-10 06:00:00-07:00    20_949.6200 | 
```

Reply:
That's the best I came up with too. 

Obviously, make sure to use pacific time when doing calculation.

If not too hard, might as well update load forecast in this PR with same logic

Reply:
I don't know if the publish times are UTC or PT. The publication schedule does not specify and I would guess PT, but the "report generated" footer is in UTC.

<img width="977" alt="Screenshot 2024-04-09 at 20 19 37" src="https://github.com/kmax12/gridstatus/assets/17484616/7413695d-33e0-4a4c-9315-e4c6325b87dc">

Reply:
There are some other reports in that list from the DAM market run that publish at 1300 and i know from other sources that those data become available at 1300 PT. so PT should be right

<img width="2127" alt="image" src="https://github.com/kmax12/gridstatus/assets/371792/663ddd30-2a6a-4cd1-b5e9-e2cbb3a7aaf0">

Reply:
Alright, I've updated the logic to avoid publish times in the future.

```python
solar_and_wind = caiso.get_solar_and_wind_forecast_dam("2024-04-08", "2024-04-12")

solar_and_wind.head()

| Interval Start            | Interval End              | Publish Time              | Location | Solar MW | Wind MW | 
|---------------------------|---------------------------|---------------------------|----------|----------|---------| 
| 2024-04-08 00:00:00-07:00 | 2024-04-08 01:00:00-07:00 | 2024-04-07 07:00:00-07:00 | CAISO    | 0.0      | 3790.96 | 
| 2024-04-08 00:00:00-07:00 | 2024-04-08 01:00:00-07:00 | 2024-04-07 07:00:00-07:00 | NP15     | 0.0      | 334.14  | 
| 2024-04-08 00:00:00-07:00 | 2024-04-08 01:00:00-07:00 | 2024-04-07 07:00:00-07:00 | SP15     | 0.0      | 3406.68 | 
| 2024-04-08 00:00:00-07:00 | 2024-04-08 01:00:00-07:00 | 2024-04-07 07:00:00-07:00 | ZP26     | 0.0      | 50.14   | 
| 2024-04-08 01:00:00-07:00 | 2024-04-08 02:00:00-07:00 | 2024-04-07 07:00:00-07:00 | CAISO    | 0.0      | 3737.54 | 

solar_and_wind.tail()

| Interval Start            | Interval End              | Publish Time              | Location | Solar MW | Wind MW | 
|---------------------------|---------------------------|---------------------------|----------|----------|---------| 
| 2024-04-11 22:00:00-07:00 | 2024-04-11 23:00:00-07:00 | 2024-04-10 07:00:00-07:00 | ZP26     | 0.0      | 5.44    | 
| 2024-04-11 23:00:00-07:00 | 2024-04-12 00:00:00-07:00 | 2024-04-10 07:00:00-07:00 | CAISO    | 0.0      | 1942.47 | 
| 2024-04-11 23:00:00-07:00 | 2024-04-12 00:00:00-07:00 | 2024-04-10 07:00:00-07:00 | NP15     | 0.0      | 940.0   | 
| 2024-04-11 23:00:00-07:00 | 2024-04-12 00:00:00-07:00 | 2024-04-10 07:00:00-07:00 | SP15     | 0.0      | 998.61  | 
| 2024-04-11 23:00:00-07:00 | 2024-04-12 00:00:00-07:00 | 2024-04-10 07:00:00-07:00 | ZP26     | 0.0      | 3.86    | 
```

Reply:
Here's load forecast:

```python
load_forecast = caiso.get_load_forecast("2024-04-08", "2024-04-12")

load_forecast.head()

| Publish Time              | Time                      | Interval Start            | Interval End              | Load Forecast | 
|---------------------------|---------------------------|---------------------------|---------------------------|---------------| 
| 2024-04-07 09:10:00-07:00 | 2024-04-08 00:00:00-07:00 | 2024-04-08 00:00:00-07:00 | 2024-04-08 01:00:00-07:00 | 19935.03      | 
| 2024-04-07 09:10:00-07:00 | 2024-04-08 01:00:00-07:00 | 2024-04-08 01:00:00-07:00 | 2024-04-08 02:00:00-07:00 | 19406.79      | 
| 2024-04-07 09:10:00-07:00 | 2024-04-08 02:00:00-07:00 | 2024-04-08 02:00:00-07:00 | 2024-04-08 03:00:00-07:00 | 19037.25      | 
| 2024-04-07 09:10:00-07:00 | 2024-04-08 03:00:00-07:00 | 2024-04-08 03:00:00-07:00 | 2024-04-08 04:00:00-07:00 | 18975.29      | 
| 2024-04-07 09:10:00-07:00 | 2024-04-08 04:00:00-07:00 | 2024-04-08 04:00:00-07:00 | 2024-04-08 05:00:00-07:00 | 19560.64      | 


load_forecast.tail()

| Publish Time              | Time                      | Interval Start            | Interval End              | Load Forecast | 
|---------------------------|---------------------------|---------------------------|---------------------------|---------------| 
| 2024-04-09 09:10:00-07:00 | 2024-04-10 19:00:00-07:00 | 2024-04-10 19:00:00-07:00 | 2024-04-10 20:00:00-07:00 | 25989.39      | 
| 2024-04-09 09:10:00-07:00 | 2024-04-10 20:00:00-07:00 | 2024-04-10 20:00:00-07:00 | 2024-04-10 21:00:00-07:00 | 26013.98      | 
| 2024-04-09 09:10:00-07:00 | 2024-04-10 21:00:00-07:00 | 2024-04-10 21:00:00-07:00 | 2024-04-10 22:00:00-07:00 | 25062.53      | 
| 2024-04-09 09:10:00-07:00 | 2024-04-10 22:00:00-07:00 | 2024-04-10 22:00:00-07:00 | 2024-04-10 23:00:00-07:00 | 23400.33      | 
| 2024-04-09 09:10:00-07:00 | 2024-04-10 23:00:00-07:00 | 2024-04-10 23:00:00-07:00 | 2024-04-11 00:00:00-07:00 | 21689.45      | 
```